File: src/f1lab/corners.py
"""Corner detection from a speed-vs-distance telemetry trace.

A corner is a prominent local minimum of the smoothed speed trace (the apex),
paired with the point where braking for it started (the local speed maximum
just before). The numpy code here is the reference implementation; the C++
core (``cpp/src/corners.cpp``) mirrors it operation for operation — including
the prefix-sum smoothing — so both backends return *identical* indices, which
the parity tests assert exactly.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np
from numpy.typing import NDArray

from f1lab import native

# FastF1 car data arrives at ~4 Hz (roughly one sample per 18 m at speed), so
# a 3-sample window already averages over ~55 m of track; anything wider
# starts to blur genuinely separate corners (e.g. Silverstone's Village/Loop
# complex) into one. Noise robustness comes from the prominence walk, not
# from heavy smoothing.
DEFAULT_SMOOTH_WINDOW = 3
DEFAULT_MIN_DROP_KMH = 15.0
DEFAULT_MIN_SEPARATION_M = 50.0


@dataclass(frozen=True)
class Corners:
    """Detected corners, in track order.

    Both arrays have the same length and hold indices into the caller's
    original telemetry — not values — so any channel can be looked up at those
    samples: ``distance_m[corners.apex_indices]`` gives the apex distances,
    ``speed_kmh[corners.brake_indices]`` the speeds at the braking points.
    """

    apex_indices: NDArray[np.int64]
    brake_indices: NDArray[np.int64]

    def __len__(self) -> int:
        """Number of corners detected."""
        return int(self.apex_indices.size)


def smooth(values: NDArray[np.float64], window: int) -> NDArray[np.float64]:
    """Centered moving average; the window shrinks at the array edges."""
    if window < 1 or window % 2 == 0:
        raise ValueError("smooth_window must be an odd integer >= 1")
    n = values.size
    half = window // 2
    # Prefix-sum formulation: same operation order as the C++ core, so both
    # backends round identically.
    prefix = np.concatenate(([0.0], np.cumsum(values)))
    idx = np.arange(n)
    lo = np.maximum(idx - half, 0)
    hi = np.minimum(idx + half, n - 1)
    return np.asarray((prefix[hi + 1] - prefix[lo]) / (hi - lo + 1), dtype=np.float64)


def detect_corners(
    distance_m: NDArray[np.float64],
    speed_kmh: NDArray[np.float64],
    *,
    smooth_window: int = DEFAULT_SMOOTH_WINDOW,
    min_drop_kmh: float = DEFAULT_MIN_DROP_KMH,
    min_separation_m: float = DEFAULT_MIN_SEPARATION_M,
    engine: str | None = None,
) -> Corners:
    """Find apex and braking points in a single lap's speed trace.

    A local minimum counts as an apex when, walking outwards in both
    directions, the smoothed speed rises by at least ``min_drop_kmh`` before
    any sample dips below the minimum itself (a deeper nearby minimum shadows
    it). Apexes closer together than ``min_separation_m`` merge into the
    slower one.
    """
    if distance_m.shape != speed_kmh.shape:
        raise ValueError("distance and speed arrays must have the same shape")
    if distance_m.size == 0:
        raise ValueError("telemetry arrays must not be empty")
    if smooth_window < 1 or smooth_window % 2 == 0:
        raise ValueError("smooth_window must be an odd integer >= 1")
    if min_drop_kmh <= 0.0:
        raise ValueError("min_drop_kmh must be positive")
    if min_separation_m < 0.0:
        raise ValueError("min_separation_m must be non-negative")
    if native.resolve_engine(engine) == "cpp":
        apex, brake = native.detect_corners(
            distance_m, speed_kmh, smooth_window, min_drop_kmh, min_separation_m
        )
        return Corners(apex_indices=apex, brake_indices=brake)
    return _detect_corners_numpy(
        distance_m, speed_kmh, smooth_window, min_drop_kmh, min_separation_m
    )


def _detect_corners_numpy(
    distance_m: NDArray[np.float64],
    speed_kmh: NDArray[np.float64],
    smooth_window: int,
    min_drop_kmh: float,
    min_separation_m: float,
) -> Corners:
    """Reference implementation of :func:`detect_corners`, in five steps.

    Inputs are assumed already validated by the caller. The prominence and
    braking-point walks are data-dependent loops that do not vectorise, which
    is precisely why the C++ twin is roughly 100x faster here — see
    ``python -m f1lab.bench``.
    """
    n = speed_kmh.size
    empty = np.array([], dtype=np.int64)
    if n < 3:
        return Corners(apex_indices=empty, brake_indices=empty)
    smoothed = smooth(speed_kmh, smooth_window)

    # Candidate apexes: local minima of the smoothed trace. `<=` on the left
    # and `<` on the right picks the last sample of a flat valley floor.
    interior = np.arange(1, n - 1)
    is_minimum = (smoothed[1:-1] <= smoothed[:-2]) & (smoothed[1:-1] < smoothed[2:])
    candidates = interior[is_minimum]

    # Prominence: walk outwards from the candidate in both directions. The
    # candidate is a corner if the speed rises by at least min_drop on each
    # side before any sample dips below the candidate itself (in which case a
    # deeper minimum nearby shadows this one). Robust to small wiggles,
    # unlike comparing against neighbouring candidates.
    accepted: list[int] = []
    for candidate in candidates:
        i = int(candidate)
        rises_left = False
        for k in range(i - 1, -1, -1):
            if smoothed[k] < smoothed[i]:
                break
            if smoothed[k] - smoothed[i] >= min_drop_kmh:
                rises_left = True
                break
        rises_right = False
        for k in range(i + 1, n):
            if smoothed[k] < smoothed[i]:
                break
            if smoothed[k] - smoothed[i] >= min_drop_kmh:
                rises_right = True
                break
        if rises_left and rises_right:
            accepted.append(i)

    # Merge apexes closer than min_separation_m, keeping the slower one.
    kept: list[int] = []
    for i in accepted:
        if kept and distance_m[i] - distance_m[kept[-1]] < min_separation_m:
            if smoothed[i] < smoothed[kept[-1]]:
                kept[-1] = i
        else:
            kept.append(i)

    # Braking point: walk back from the apex tracking the running maximum;
    # stop once the trace falls min_drop below that maximum (we have crossed
    # the peak of the approach and are descending into the previous corner).
    brakes: list[int] = []
    for apex in kept:
        best = apex
        k = apex
        while k > 0:
            k -= 1
            if smoothed[k] > smoothed[best]:
                best = k
            if smoothed[best] - smoothed[k] >= min_drop_kmh:
                break
        brakes.append(best)

    return Corners(
        apex_indices=np.asarray(kept, dtype=np.int64),
        brake_indices=np.asarray(brakes, dtype=np.int64),
    )

File: cpp/src/corners.cpp
#include "corners.hpp"

#include <algorithm>
#include <stdexcept>

namespace f1lab {

std::vector<double> smooth(std::span<const double> values, std::size_t window) {
    if (window < 1 || window % 2 == 0) {
        throw std::invalid_argument("smooth_window must be an odd integer >= 1");
    }
    const std::size_t n = values.size();
    // Prefix sums turn every window average into two lookups, so the whole
    // pass is O(n) rather than O(n * window). The sequential accumulation
    // also matches np.cumsum's operation order, which is what lets the two
    // engines agree bit for bit — float addition is not associative, so the
    // order is part of the contract, not an implementation detail.
    std::vector<double> prefix(n + 1, 0.0);
    for (std::size_t i = 0; i < n; ++i) {
        prefix[i + 1] = prefix[i] + values[i];
    }
    const std::size_t half = window / 2;
    std::vector<double> out(n);
    for (std::size_t i = 0; i < n; ++i) {
        // Clip the window to the array instead of padding: near the edges it
        // simply averages fewer samples. `i > half` guards the subtraction —
        // std::size_t is unsigned, so i - half would wrap around.
        const std::size_t lo = i > half ? i - half : 0;
        const std::size_t hi = std::min(i + half, n > 0 ? n - 1 : 0);
        out[i] = (prefix[hi + 1] - prefix[lo]) / static_cast<double>(hi - lo + 1);
    }
    return out;
}

CornerResult detect_corners(std::span<const double> distance_m, std::span<const double> speed_kmh,
                            std::size_t smooth_window, double min_drop_kmh,
                            double min_separation_m) {
    if (distance_m.size() != speed_kmh.size()) {
        throw std::invalid_argument("distance and speed arrays must have the same shape");
    }
    if (distance_m.empty()) {
        throw std::invalid_argument("telemetry arrays must not be empty");
    }
    if (min_drop_kmh <= 0.0) {
        throw std::invalid_argument("min_drop_kmh must be positive");
    }
    if (min_separation_m < 0.0) {
        throw std::invalid_argument("min_separation_m must be non-negative");
    }

    CornerResult result;
    const std::size_t n = speed_kmh.size();
    if (n < 3) {
        return result;  // Too short to contain an interior local minimum.
    }
    // Step 1: smooth. Every decision below reads `s`, but the indices we
    // return address the caller's original arrays.
    const std::vector<double> s = smooth(speed_kmh, smooth_window);

    // Step 2: collect local minima as corner candidates. The asymmetry —
    // `<=` on the left, `<` on the right — makes a flat valley floor yield
    // exactly one candidate (its last sample) instead of one per sample.
    std::vector<std::size_t> candidates;
    for (std::size_t i = 1; i + 1 < n; ++i) {
        if (s[i] <= s[i - 1] && s[i] < s[i + 1]) {
            candidates.push_back(i);
        }
    }

    // Step 3: keep only prominent candidates. Walk outwards from each one in
    // both directions: a rise of min_drop_kmh before any sample dips below
    // the candidate means this really is the bottom of its own corner. Hitting
    // a lower sample first means a deeper minimum nearby shadows it — that
    // deeper one is the real apex, and this candidate is just a wiggle on the
    // way into it.
    std::vector<std::size_t> accepted;
    for (const std::size_t i : candidates) {
        bool rises_left = false;
        // `k-- > 0` tests then decrements, so the body sees i-1 ... 0 and the
        // loop still terminates on unsigned k (a plain `k >= 0` never would).
        for (std::size_t k = i; k-- > 0;) {
            if (s[k] < s[i]) {
                break;
            }
            if (s[k] - s[i] >= min_drop_kmh) {
                rises_left = true;
                break;
            }
        }
        bool rises_right = false;
        for (std::size_t k = i + 1; k < n; ++k) {
            if (s[k] < s[i]) {
                break;
            }
            if (s[k] - s[i] >= min_drop_kmh) {
                rises_right = true;
                break;
            }
        }
        if (rises_left && rises_right) {
            accepted.push_back(i);
        }
    }

    // Step 4: collapse apexes closer than min_separation_m into the slower
    // one, so a chicane counts as a single corner. Candidates arrive in track
    // order, so comparing against the last kept apex is enough.
    std::vector<std::size_t> kept;
    for (const std::size_t i : accepted) {
        if (!kept.empty() && distance_m[i] - distance_m[kept.back()] < min_separation_m) {
            if (s[i] < s[kept.back()]) {
                kept.back() = i;
            }
        } else {
            kept.push_back(i);
        }
    }

    // Step 5: find where braking began. Walk back from the apex tracking the
    // highest speed seen so far; once the trace has fallen min_drop_kmh below
    // that peak we have crossed it and are descending into the previous
    // corner, so the peak itself is the braking point. Tracking a running
    // maximum (rather than stopping at the first sample that is not higher
    // than its neighbour) keeps this stable on a noisy approach.
    for (const std::size_t apex : kept) {
        std::size_t best = apex;
        std::size_t k = apex;
        while (k > 0) {
            --k;
            if (s[k] > s[best]) {
                best = k;
            }
            if (s[best] - s[k] >= min_drop_kmh) {
                break;
            }
        }
        result.apex_indices.push_back(static_cast<std::int64_t>(apex));
        result.brake_indices.push_back(static_cast<std::int64_t>(best));
    }
    return result;
}

}  // namespace f1lab
